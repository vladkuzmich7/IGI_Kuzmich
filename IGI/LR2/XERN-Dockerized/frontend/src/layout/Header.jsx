import React from 'react'
import { Nav, Navbar, Button, DropdownButton, Dropdown, Image } from 'react-bootstrap'
import { Redirect, withRouter, NavLink, Link } from 'react-router-dom'

import {connect} from 'react-redux'
import {sources, StyleSettings} from '../settings/config';
import {LoggedOut} from '../actions/userAuth';
import SessionButton from '../components/buttons/SessionButton';
import LoginButton from '../components/buttons/LoginButton';
import styled from 'styled-components';



const StyledHeader = styled.nav`


position: fixed;
background-color: var(--theme-color);
color: var(--background-app);
top: 0;
width: 100%;
display: flex;
height: var(--width-header-sidebar);
z-index: 3;

img{
  width: 8.5rem;
  height: 5rem;
  margin-left: 4rem;
  margin-top: -.5rem;  
}

.header-title {
  margin: auto 1rem;
  text-decoration: none;
  color: var(--background-app);
}

.buttons {
  margin: .7rem 1rem .3rem auto;
  
  
  button {
    background: transparent;
    color: var(--background-app);
  }

  svg{
    width: 1.5rem;
    height: 1.5rem;
    margin:  0 .5rem;
    color: var(--background-app);
  }
}

.no-buttons {
  button {
    display: none;
  }
}

@media screen and (max-width: ${StyleSettings.MaxDisplayMobile}) {
  
.header-title, .login-text, .session-username{
  display: none;
}

img {
  width: 8.5rem;
  height: 5rem;
  margin-left: 0rem;
  padding: .5rem; 
}

}
`


const Header = ({username, status, LoggedOut, history}) => {
  
  //if(history.location.pathname ==='/login'){return null}
  
  return (
    
    <StyledHeader>
      {/* {console.log(history.location.pathname, 'history on HEADER')} */}
      <img src={sources.logo} alt="MERN Boilerplate"/>
      <NavLink to="/" className="header-title">
            Boilerplate
      </NavLink>      
      <span className={history.location.pathname !== '/login' ? 'buttons': 'no-buttons'}>
        {
          username !== 'anonymous' && status === 'logged' ? 
            <SessionButton 
              username={username}
              LoggedOut={LoggedOut}
              status={status}
            />        
            : 
            <LoginButton/>
        }
      </span>    
    </StyledHeader>
  )
}

const mapStateToProps = (state) => ({
  username: state.user.username,
  status: state.user.status,

})

const mapDispatchToProps = dispatch => ({
  LoggedOut: () => dispatch(LoggedOut())
})

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Header))
