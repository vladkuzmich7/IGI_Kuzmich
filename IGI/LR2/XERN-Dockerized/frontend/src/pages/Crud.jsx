import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import Title from '../components/Title'
import { withRouter } from 'react-router-dom';
import MainTable from '../components/tables/MainTable'
import CrudData from '../components/tables/CrudData'
import {ModalData, useModal} from '../components/ModalData'
import { sortData } from '../selectors/sortData';
import FormData from '../components/forms/FormData';
import {setSort} from '../actions/sort'
import fetchCrudApi from '../actions/fetchCrudApi'
import {sources, display, StyleSettings} from '../settings/config'
import CreateButton from '../components/buttons/CreateButton';
import Loading from '../components/Loading';   
import styled from 'styled-components'


const StyledCrud = styled.div`

    .create-button {
        margin: 0% 45%;
    }


@media screen and (max-width: ${StyleSettings.MaxDisplayMobile}) {

    h1 {
        font-size: 1.5rem;
    }

    .create-button {
        margin: 5% 35%;
    }

    .main-table {

        font-size: .75rem;

        table {
            margin: 0;
        }
    }
}
`

// export const crudNamesCols = [
//     'Name',
//     'Price',
//     'Quantity',
//     'Description',
// ]


const Crud = ({result, fetchCrudApi, loading, sortKey, sortDirection, setSort, auth, history}) => {

    const {show, handleShow, handleClose, data, setData} = useModal()
    const [action, setAction] = useState('list')
    const crudColumns = {
        names: display.crudNamesCols,
        sortKey,
        sortDirection
}
        
    useEffect(() => {
        
        fetchCrudApi(sources.dataAdmin,'get',null)
        //console.log(action, "Form action")
        }, [])

        

    return (

        <StyledCrud>
            {loading === true ? <Loading message={'Loading'}/>
                :
            <div>            
                <Title content={'CRUD Example'}/>
            <div className="create-button">
                <CreateButton
                    handleShow={handleShow}
                    setAction={setAction}
                /> 
            </div>
            <div className="main-table">
                <MainTable
                    columns={crudColumns}
                    handleShow={handleShow}
                    loading={loading}
                    setAction={setAction}
                    setSort={setSort}
                >   
                    <CrudData 
                        setData={setData}
                        fetchCrudApi={fetchCrudApi}
                        result={result}
                        colLength={crudColumns.names.length}
                        handleShow={handleShow}
                        setAction={setAction}
                    />
                </MainTable>
            </div>
            <ModalData show={show} handleClose={handleClose} title={action}>
                {   action === 'create' ?
                    <FormData data={{ name: '', price: '', quantity: '', description: ''}}
                        fetchCrudApi={fetchCrudApi}
                        action={'post'}
                        handleClose={handleClose}
                    />
                :    
                    action === 'update' && data ?
                    <FormData data={{                         
                        name: data.name, 
                        price: data.price, 
                        quantity: data.quantity, 
                        description: data.description}}
                        fetchCrudApi={fetchCrudApi}
                        handleClose={handleClose}
                        action={'patch'}
                        idparam={data._id}
                    />
                :                                
                    action === 'list' && data ? 
                <>
                    <p>{data.name}</p>
                    <p>{data.price}</p>
                </>    
                :
                    <p>NO DATA</p>
                }
            </ModalData>                    
        </div>
        }
    </StyledCrud>
    )
}

const mapStateToProps = (state) => ({
    result: sortData(state.example), 
    sortKey: state.example.sortKey,
    sortDirection: state.example.sortDirection,
    loading: state.example.loading, 
    auth: state.user
})

const mapDispatchToProps = dispatch => ({
    fetchCrudApi: (url, method, data) => dispatch(fetchCrudApi(url, method, data)),
    setSort: sortkey => dispatch(setSort(sortkey))
})

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Crud))
