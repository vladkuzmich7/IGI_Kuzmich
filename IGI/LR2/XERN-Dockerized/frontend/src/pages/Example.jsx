import React, {useState, useEffect} from 'react'
import {connect} from 'react-redux';
import { withRouter } from 'react-router-dom';
import {Card, Button, Spinner} from 'react-bootstrap'
import {sources, StyleSettings} from '../settings/config';
import {FaCity, FaCloud, FaSun, FaTemperatureHigh, FaTemperatureLow,FaWind, FaCloudRain, FaSnowflake,
    FaMoon, FaRedditAlien
} 

from 'react-icons/fa'
import {IoIosThunderstorm} from 'react-icons/io'
import {FiCloudDrizzle} from 'react-icons/fi'

import styled from 'styled-components'


const StyledExample = styled.div`

    
    h1 {
        font-size: 2rem;
        text-align: center;
        margin: 2rem auto;
    }


    .list-card-socket, .list-card-scrap {
        display: flex;
        flex-wrap: wrap;
        padding: 0;
        justify-content: center;
        text-align: center;
    }

    .list-card-socket > div , .list-card-scrap > div{
        box-shadow: 0 4px 8px 0 rgba(0,0,0,0.3);
    }


    .card-example-weather, .card-example-empty, .card-example-scrap  {
        width: 25%;
        margin: 2rem;
        background: var(--card-background);
        transition: all 0.75s ease;
    }

    .weather-header{
        display: flex;
        flex-direction: row;
    
        span{
            margin:2rem;
            width: 40%;
            color: black;
        }
    
        h2{
            margin: 2rem 2rem .5rem 2rem;
            width: 60%;
            color: black;

            svg{
                margin-left: 2rem;
                color: black;
            }
        }
    }

    .weather-body{
        display: flex;
        flex-direction: row;
    
        span{
            margin:.1rem;
            width: 65%
        }

        svg {
            width: 4rem;
            height: 4rem;
            color: black;
        }
    
        p {
            margin: 2rem;
            width: 35%;
            color: black;
        }
    }

    button {
        background: darkblue;
        display: flex;
        justify-content: center;
        margin: 0 auto;
        width: 9rem;
    }


    a {
        text-decoration: none;
    }

    .scrap-header {
        background: var(--theme-color);
        color: var(--background-app);
        padding: 1.5rem;
        height: 5rem;
    }

    .scrap-body{
    
        margin: 2rem auto auto auto;
        height: 100%;
        position: relative;
    
        p {
            z-index: 2;
            position: relative;
            margin: 2rem 1rem;
            color: black;
        }

        svg {
            position: absolute;
            color: rgba(248,102,11,0.2);
            top: 0;       
            margin: 1rem 3rem 1rem 3rem;
            z-index: 1;
            width: 3rem;
            height: 3rem;
        }
    }


    .card-example-scrap:hover{
        background: linear-gradient(to bottom, rgba(82,82,82,1) 0%, rgba(46,46,46,1) 1%, rgba(97,97,97,0.99) 12%, rgba(110,110,110,0.88) 25%, rgba(77,77,77,0.75) 39%, rgba(20,20,20,0.61) 73%, rgba(31,31,31,0.27) 91%, rgba(46,46,46,0.09) 98%, rgba(20,20,20,0.02) 100%);
        transform: scale(1.1);
        p{
            color: white;
        }
    }


@media screen and (max-width: ${StyleSettings.MaxDisplayMobile}) {


    h1 {
        font-size: 1.5rem;
        text-align: center;
        margin: 2rem auto;
    }

    .card-example-weather, .card-example-empty, .card-example-scrap {
        width: 45%;
        margin: 2rem .5rem;    
    }

    .weather-header{
    
        span{
            margin: 1rem .5rem .5rem;
            width: 40%
        }
    
        h2  {
            margin: 2rem 2rem .5rem .5rem;
            width: 60%;
            font-size: 1.5rem;

            svg {
                margin-left: 2rem;
            }
        }
    }

    .weather-body{

        p{
            margin: 2rem 2rem 2rem 1rem;
            width: 35%;
        }
}

    .scrap-body{
    
        svg {
            margin: 1rem;
        }

        p {
            font-size: .75rem
        }

    }

    .scrap-header {
        background: black;
        color: whitesmoke;
        padding: 1.5rem;
        height: 5rem;
    }
} 

`

const Example = ({weather, scrap}) => {

    const[scrapStatus, setscrapStatus] = useState('Inactive')
    const scrapNews = (url) => {

        fetch(url)
        .then(response => response.json())
        .then(response => setscrapStatus(response.message))
        .catch(error => setscrapStatus(error))
    }
    
    
    useEffect(() => {
        
        scrapNews(sources.scrapUrl)
        setscrapStatus('scraping')
    }, [])

    //console.log(scrapStatus, "SCRAP STATUS")
    //console.log(weather, "DATA WEATHER");
    return (
        <StyledExample>
            <h1>Socket Example</h1>
            <div className="list-card-socket">
                
                {weather && weather.length > 0 ?
                weather.map((data) => (
                <div className="card-example-weather" key={data.name}>
                    <div className="weather-header">
                        <span>{data.name}</span>
                        <h2>{data.main.temp}
                            {                        
                                data.main.temp > 15 ?
                                    <FaTemperatureHigh/>
                                    :
                                    <FaTemperatureLow/>
                            }                        
                        </h2>                        
                    </div>
                    <div className="weather-body">
                        <span>
                        {
                            {
                                'Clear': <FaSun/>,
                                'Clouds': <FaCloud/>,
                                'Snow': <FaSnowflake/>,
                                'Rain': <FaCloudRain/>,
                                'Drizzle': <FiCloudDrizzle/>,
                                'Thunderstorm': <IoIosThunderstorm/>
                            }[data.weather[0].main]
                            
                        } 
                        </span>                               
                        <p>{data.wind.speed}m/s</p>                        
                    </div>
                </div>
                ))
                :
                <div className="card-example-empty">
                    <div className="card-example-header">Weather info</div>
                    <div className="card-example-body">NO DATA</div>
                </div>                
                }
            </div>
            <h1>Scrap Example</h1>            
                <p>
                    <Button onClick={() => 
                        {scrapNews(sources.scrapUrl)
                        setscrapStatus('scraping')}
                    }>
                        Scrap Now
                    </Button>
                </p>            
            <div className="list-card-scrap">
                
                {scrap && scrap.length > 0 ?
            
                scrap.map((data) => (
                <div className="card-example-scrap" key={data.link}>
                    <a href={data.link}>
                    <div className="scrap-header">Reddit News</div>
                    <div className="scrap-body">                        
                        <p>{data.title}</p>                        
                            <FaRedditAlien/>                        
                    </div>
                    </a>
                </div>
                ))
                :
                scrapStatus === 'scraping' ?
                <div className="card-example-loading">
                    <div className="card-exmaple-header">Scrap Info</div>
                    <div className="card-exmaple-body"> Scraping . . .<Spinner animation="border" variant="dark" /> </div>
                </div>
                :    
                <div className="card-example-empty">
                    <div className="card-exmaple-header">Scrap Info</div>
                    <div className="card-exmaple-body"> NO DATA</div>
                </div>
                }
            </div>
        </StyledExample>
    )
}

const mapStateToProps = state => {
    return {
        weather: state.socketdata.wsData.job,
        scrap: state.socketdata.wsData.otherjob
    }
}

export default connect(mapStateToProps, null)(withRouter(Example))