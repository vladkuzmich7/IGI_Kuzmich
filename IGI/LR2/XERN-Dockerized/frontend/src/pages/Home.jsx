import React, {useEffect, useState, useCallback} from 'react'
import {connect} from 'react-redux'
import { withRouter, Link, NavLink } from 'react-router-dom';
import fetchCrudApi from '../actions/fetchCrudApi'
//import {ConnectWS} from '../actions/wsocket';
import {exampleSelector} from '../selectors/exampleSelector'
import socketIOClient from "socket.io-client";
import {sources, StyleSettings} from '../settings/config'
import Result from '../components/Result'
import { Card, Spinner, Button} from 'react-bootstrap'
import CardTemplate from '../components/CardTemplate';
import styled from 'styled-components';
import {FaNotesMedical, FaBusinessTime, FaThList, FaUsersCog, FaGlobeAmericas, FaNetworkWired, FaCubes} from 'react-icons/fa'


const StyledHome = styled.div`


h1{
    margin:  2rem auto;
    text-align: center;
}

.card-list {
    display: flex;
}

h2 {
    padding: 1rem 0 0 1rem;
}

@media screen and (max-width: ${StyleSettings.MaxDisplayMobile}) {

    h1{
    margin:  1rem auto;
    font-size: 2rem;
}

    .card-list {
    flex-wrap: wrap;
    margin: 0;
}
    h2{
        font-size: 1rem;
    }
}
`

const Home = ({fetchCrudApi, result, websocket, history}) => {


    //const [websocket, setWsStatus] = useState('checking...')
    const [healthCheck, setHealthCheck] = useState('checking...')
    const [homeLoading, sethomeLoading] = useState(false)
    const [backendApiData, setBackendData] = useState('')
    const checkApiHealth = (Url) => {

        if(Url === sources.HealthEndpoint){
            fetch(Url)
            .then((response) => {
                setHealthCheck(response.statusText)
                // console.log(healthCheck, "HEALTHCHECK STATEzzzz");
            })
        
            .catch((error) => {setBackendData('error')})
        }
        if(Url === sources.checkApiUrl){
            sethomeLoading(true)
            fetch(Url)
            .then(response => response.json())
            .then(jsonres => {    
                console.log(jsonres, "api response");
                sethomeLoading(false)
                setBackendData(jsonres)
                console.log(backendApiData, "backend api state");
            })
            .catch((error) => {
                sethomeLoading(false)
                setBackendData('error')            
        })
            
            }
    }

    useEffect(() => {   
        //checkWebsocket(sources.WSocket)
//        ConnectWS(sources.WSocket)
        checkApiHealth(sources.HealthEndpoint)
    }, [])

    return (
        <StyledHome>
            <h1>Services Check</h1>            
            <div className='card-list'>                
                <CardTemplate title={'Endpoint Health'} 
                customStyle={ healthCheck === 'OK' ? 'healthy' : 'error'}                
                >
                        <FaNotesMedical onClick={() => checkApiHealth(sources.HealthEndpoint)}/>
                        <h2>{healthCheck}</h2>
                </CardTemplate>
                <CardTemplate title={'Scheduled Tasks'} customStyle="normal-content">                        
                            <a href={sources.taskMonitor}>
                                <FaBusinessTime/>
                            </a>                        
                </CardTemplate>
                <CardTemplate title={'Crud Example'} customStyle="normal-content">                   
                            <Link to='/list'>
                                <FaThList />                                
                            </Link>
                </CardTemplate>
                <CardTemplate title={'Users Admin'} customStyle="normal-content">
                    <Link to='/users'>
                        <FaUsersCog/>
                    </Link>        
                </CardTemplate>
                <CardTemplate title={'Socket/Scraper Example'} customStyle="normal-content">
                    <Link to='/example'>
                        <FaGlobeAmericas/>                        
                    </Link>
                </CardTemplate>
            </div>
            <div className='card-list'>
                <CardTemplate title ={'Websocket State'} customStyle={websocket.status === 'connected' ||  websocket.status === 'message received' ? 
                                    'healthy' : 'error'}>
                        <FaNetworkWired/>                        
                        <h2>{websocket.status}</h2>
                </CardTemplate>            
                
                <CardTemplate title={'Backend Api request'} customStyle='api-request'>
                        <FaCubes onClick={() => checkApiHealth(sources.checkApiUrl)}/>
                        {    
                            !backendApiData.data  &&
                            null                    
                        }        
                        {    
                            homeLoading === true &&
                            <div >
                                <Card.Header>
                                    <Spinner animation="border" role="status" variant='dark'/>
                                    Loading
                                </Card.Header>
                            </div>
                        }
                        {
                            backendApiData.message && 
                            <Result type='Success' message={JSON.stringify(backendApiData.message, null, 2)}/>                        
                        }
                        
                        {   (backendApiData.error && backendApiData.error.message) &&
                            <Result type='Error' message={backendApiData.error.message}/>
                        }                    
                </CardTemplate>                
            </div>
        </StyledHome>
    )
}

const mapStateToProps = state => {
    return {
        result: state.example,
        websocket: state.socketdata
    }
}

export default connect(mapStateToProps, { fetchCrudApi })(withRouter(Home))
