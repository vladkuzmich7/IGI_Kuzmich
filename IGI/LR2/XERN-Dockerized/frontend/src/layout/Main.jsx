import React, {useEffect, useCallback} from 'react'
import Header from '../layout/Header'
import Sidebar from '../layout/Sidebar'
import Home from '../pages/Home'
import Crud from '../pages/Crud'
import Login from '../pages/Login';
import Example from '../pages/Example'
import Users from '../pages/Users'
import Profile from '../pages/Profile'
import Footer from '../layout/Footer'
import {ConnectWS} from '../actions/wsocket';
import {connect} from 'react-redux';
import {sources, StyleSettings} from '../settings/config';
import { createBrowserHistory } from 'history'
import { Router, Route, Switch } from 'react-router-dom'
import PrivateRoute from '../helpers/PrivateRoute';
import styled from 'styled-components'


const Wrapper = styled.div`

width: 100vw;
height: 100vh;
background: var(--background-app);
color: var(--font-color);
/*
display: grid;
grid-template-columns: auto auto;
*/

svg{
  color: var(--background-app);
}

.content {
    padding: 4rem 0rem 2rem 4rem;        
    width: 100%;
    height: 100%;
  }

  .sidebar {
    margin-top: var(--width-header-sidebar);
    width: var(--width-header-sidebar);
    background: var(--theme-color);
    position: fixed;
    z-index: 3;
    height: 100vh;
    transition: width 200ms ease;
    

  }

  .sidebar:hover{
    width: var(--width-full-sidebar);
  }

  .sidebar:hover .link-text {
    display: block;
  }


@media screen and (max-width: ${StyleSettings.MaxDisplayMobile}) {

  .sidebar {
    bottom: 0;
    width: 100vw;
    height: var(--width-header-sidebar);
    overflow: scroll;
  }
  
  .sidebar:hover {
    width: 100%;
  }

  .sidebar link.text:hover {
    display: inline;
    font-size: .5rem;
    
  }

  .content {
    padding: 4rem 0rem 0rem 0rem;
    overflow: scroll;
  }
}


`



export const hist = createBrowserHistory()


const pages = {

  normal: [
  {path: '/', component: Home},
  {path: '/login', component: Login},
  {path: '/example', component: Example},
  ],

  protected: [
    {path: '/list', component: Crud},
    {path: '/users', component: Users},
    {path: '/profile', component: Profile},
  ]
}


const Main = ({ConnectWS}) => {

  useEffect(() => {
    ConnectWS(sources.WSocket)
    }
  , [])


  return (
    
      <Router history={hist}>
        <Header />
        <Wrapper>
          <div className="sidebar">
            <Sidebar/>
          </div>
          <div className="content">
            <Switch>
              {pages.normal.map((page) => 
                <Route exact path={page.path} component={page.component} key={page.path}/>
              )}
              <PrivateRoute>
              {pages.protected.map((page) => 
                <Route exact path={page.path} component={page.component} key={page.path}/>
              )}
              </PrivateRoute>
            </Switch>
          </div>
        </Wrapper>
        <Footer />
      </Router>
    
  )
}

export default connect(null, {ConnectWS})(Main)
