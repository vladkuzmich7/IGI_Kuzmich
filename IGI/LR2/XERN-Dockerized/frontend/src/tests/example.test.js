import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import Main from '../layout/Main';
import Header from '../layout/Header'
import Home from '../pages/Home'
import Crud from '../pages/Crud'
import Login from '../pages/Login';
import Example from '../pages/Example'
import Footer from '../layout/Footer'


const mockStore = configureStore([]);

describe('Header', () => {
  let store;
  let component;
  
  beforeEach(() => {
    store = mockStore({
      myState: 'sample text',
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <Header />
      </Provider>
    );
  });
  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('should dispatch an action on button click', () => {
    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      myAction({ payload: 'sample text' })
    );
  });
   
  });


  