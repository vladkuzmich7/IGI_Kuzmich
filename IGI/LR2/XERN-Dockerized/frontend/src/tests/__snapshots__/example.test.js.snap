// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer snapshot renders 1`] = `<div />`;
